File: Shared/Protocol.swift
import Foundation

enum Button: String, Codable, Sendable { case left, right }
enum Key: String, Codable, Sendable { case left, right, esc }
enum Chord: String, Codable, Sendable { case spaceLeft, spaceRight, missionControl }
enum Command: Equatable, Sendable, Codable {
    case auth(code: String), move(dx: Double, dy: Double), click(Button), scroll(dx: Double, dy: Double), key(Key), chord(Chord)
    private enum Fields: String, CodingKey { case t, code, dx, dy, b, k }
    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: Fields.self)
        let type = try c.decode(String.self, forKey: .t)
        let fields: [String]
        switch type {
        case "auth": fields = ["t", "code"]
        case "move", "scroll": fields = ["t", "dx", "dy"]
        case "click": fields = ["t", "b"]
        case "key", "chord": fields = ["t", "k"]
        default: throw DecodingError.dataCorruptedError(forKey: .t, in: c, debugDescription: "Unknown command")
        }
        try requireFields(fields, from: decoder)
        switch type {
        case "auth": self = .auth(code: try c.decode(String.self, forKey: .code))
        case "move", "scroll":
            let x = try c.decode(Double.self, forKey: .dx), y = try c.decode(Double.self, forKey: .dy)
            guard x.isFinite && y.isFinite else { throw DecodingError.dataCorruptedError(forKey: .dx, in: c, debugDescription: "Nonfinite delta") }
            self = try c.decode(String.self, forKey: .t) == "move" ? .move(dx: x, dy: y) : .scroll(dx: x, dy: y)
        case "click": self = .click(try c.decode(Button.self, forKey: .b))
        case "key": self = .key(try c.decode(Key.self, forKey: .k))
        case "chord": self = .chord(try c.decode(Chord.self, forKey: .k))
        default: throw DecodingError.dataCorruptedError(forKey: .t, in: c, debugDescription: "Unknown command")
        }
    }
    func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: Fields.self)
        switch self {
        case .auth(let code): try c.encode("auth", forKey: .t); try c.encode(code, forKey: .code)
        case .move(let x, let y), .scroll(let x, let y):
            guard x.isFinite && y.isFinite else { throw EncodingError.invalidValue(self, .init(codingPath: encoder.codingPath, debugDescription: "Nonfinite delta")) }
            if case .move = self { try c.encode("move", forKey: .t) } else { try c.encode("scroll", forKey: .t) }
            try c.encode(x, forKey: .dx); try c.encode(y, forKey: .dy)
        case .click(let b): try c.encode("click", forKey: .t); try c.encode(b, forKey: .b)
        case .key(let k): try c.encode("key", forKey: .t); try c.encode(k, forKey: .k)
        case .chord(let k): try c.encode("chord", forKey: .t); try c.encode(k, forKey: .k)
        }
    }
}
enum Reply: Equatable, Sendable, Codable {
    case status, bye(reason: String)
    private enum Fields: String, CodingKey { case t, reason }
    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: Fields.self)
        let type = try c.decode(String.self, forKey: .t)
        let fields: [String]
        switch type {
        case "status": fields = ["t"]
        case "bye": fields = ["t", "reason"]
        default: throw DecodingError.dataCorruptedError(forKey: .t, in: c, debugDescription: "Unknown reply")
        }
        try requireFields(fields, from: decoder)
        switch type {
        case "status": self = .status
        case "bye": self = .bye(reason: try c.decode(String.self, forKey: .reason))
        default: throw DecodingError.dataCorruptedError(forKey: .t, in: c, debugDescription: "Unknown reply")
        }
    }
    func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: Fields.self)
        switch self {
        case .status: try c.encode("status", forKey: .t)
        case .bye(let reason): try c.encode("bye", forKey: .t); try c.encode(reason, forKey: .reason)
        }
    }
}

private struct WireField: CodingKey {
    let stringValue: String
    var intValue: Int? { nil }
    init?(stringValue: String) { self.stringValue = stringValue }
    init?(intValue: Int) { return nil }
}

private func requireFields(_ expected: [String], from decoder: Decoder) throws {
    let fields = try decoder.container(keyedBy: WireField.self)
    guard Set(fields.allKeys.map(\.stringValue)) == Set(expected) else {
        throw DecodingError.dataCorrupted(.init(codingPath: decoder.codingPath, debugDescription: "Unexpected frame fields"))
    }
}

enum WireProtocol {
    static func decodeCommand(_ data: Data) throws -> Command {
        try requireUniqueKeys(data)
        return try JSONDecoder().decode(Command.self, from: data)
    }

    static func decodeReply(_ data: Data) throws -> Reply {
        try requireUniqueKeys(data)
        return try JSONDecoder().decode(Reply.self, from: data)
    }

    // Codable containers collapse duplicate keys before init(from:) can inspect them.
    private static func requireUniqueKeys(_ data: Data) throws {
        guard String(data: data, encoding: .utf8) != nil else { throw invalidFrame() }
        let bytes = Array(data)
        var index = 0, depth = 0
        var expectingKey = true
        var keys = Set<String>()
        while index < bytes.count {
            let byte = bytes[index]
            if byte == 34 {
                let start = index
                index += 1
                while index < bytes.count {
                    if bytes[index] == 92 { index += 2; continue }
                    if bytes[index] == 34 { break }
                    index += 1
                }
                guard index < bytes.count else { throw invalidFrame() }
                if depth == 1 && expectingKey {
                    let key = try JSONDecoder().decode(String.self, from: Data(bytes[start...index]))
                    guard keys.insert(key).inserted else { throw invalidFrame() }
                }
            } else if byte == 123 || byte == 91 {
                depth += 1
            } else if byte == 125 || byte == 93 {
                if depth == 1 && expectingKey && !keys.isEmpty { throw invalidFrame() }
                depth -= 1
            } else if depth == 1 && byte == 44 {
                expectingKey = true
            } else if depth == 1 && byte == 58 {
                expectingKey = false
            }
            index += 1
        }
    }

    private static func invalidFrame() -> DecodingError {
        .dataCorrupted(.init(codingPath: [], debugDescription: "Invalid JSON frame or duplicate field"))
    }
}

File: host/internal/protocol/protocol.go
package protocol

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"math"
	"strings"
	"unicode/utf8"
)

const MaxAbsMove = 400

// MaxFrameBytes is the maximum attribute value carried by one BLE write.
const MaxFrameBytes = 512

// MinimumCommandBytes accommodates canonical finite-double movement frames.
const MinimumCommandBytes = 80

var (
	ErrInvalidFrame = errors.New("invalid frame")
	ErrNonfinite    = errors.New("nonfinite delta")
)

type Command interface{ command() }

type Auth struct{ Code string }
type Move struct{ Dx, Dy float64 }
type Click struct{ Button Button }
type Scroll struct{ Dx, Dy float64 }
type KeyPress struct{ Key Key }
type ChordPress struct{ Chord Chord }

func (Auth) command()       {}
func (Move) command()       {}
func (Click) command()      {}
func (Scroll) command()     {}
func (KeyPress) command()   {}
func (ChordPress) command() {}

type Button string
type Key string
type Chord string

const (
	ButtonLeft  Button = "left"
	ButtonRight Button = "right"
	KeyLeft     Key    = "left"
	KeyRight    Key    = "right"
	KeyEsc      Key    = "esc"
	SpaceLeft   Chord  = "spaceLeft"
	SpaceRight  Chord  = "spaceRight"
	MissionCtrl Chord  = "missionControl"
)

type Reply interface{ reply() }

type Status struct{}
type Bye struct{ Reason string }

func (Status) reply() {}
func (Bye) reply()    {}

const (
	ReasonKicked    = "kicked"
	ReasonDisplaced = "displaced"
	ReasonBadAuth   = "badauth"
)

func ParseCommand(data []byte) (Command, error) {
	raw, err := parseObject(data)
	if err != nil {
		return nil, err
	}
	t, _ := raw["t"].(string)
	fields := map[string][]string{
		"auth": {"t", "code"}, "move": {"t", "dx", "dy"}, "scroll": {"t", "dx", "dy"},
		"click": {"t", "b"}, "key": {"t", "k"}, "chord": {"t", "k"},
	}
	if !hasFields(raw, fields[t]) {
		return nil, ErrInvalidFrame
	}
	switch t {
	case "auth":
		code, ok := raw["code"].(string)
		if !ok {
			return nil, ErrInvalidFrame
		}
		return Auth{Code: code}, nil
	case "move", "scroll":
		dx, err := number(raw["dx"])
		if err != nil {
			return nil, err
		}
		dy, err := number(raw["dy"])
		if err != nil {
			return nil, err
		}
		if t == "move" {
			return Move{Dx: dx, Dy: dy}, nil
		}
		return Scroll{Dx: dx, Dy: dy}, nil
	case "click":
		b, ok := raw["b"].(string)
		if !ok || (b != string(ButtonLeft) && b != string(ButtonRight)) {
			return nil, ErrInvalidFrame
		}
		return Click{Button: Button(b)}, nil
	case "key":
		k, ok := raw["k"].(string)
		if !ok || (k != string(KeyLeft) && k != string(KeyRight) && k != string(KeyEsc)) {
			return nil, ErrInvalidFrame
		}
		return KeyPress{Key: Key(k)}, nil
	case "chord":
		k, ok := raw["k"].(string)
		if !ok || (k != string(SpaceLeft) && k != string(SpaceRight) && k != string(MissionCtrl)) {
			return nil, ErrInvalidFrame
		}
		return ChordPress{Chord: Chord(k)}, nil
	default:
		return nil, ErrInvalidFrame
	}
}

func ParseReply(data []byte) (Reply, error) {
	raw, err := parseObject(data)
	if err != nil {
		return nil, err
	}
	t, _ := raw["t"].(string)
	fields := map[string][]string{"status": {"t"}, "bye": {"t", "reason"}}
	if !hasFields(raw, fields[t]) {
		return nil, ErrInvalidFrame
	}
	switch t {
	case "status":
		return Status{}, nil
	case "bye":
		reason, ok := raw["reason"].(string)
		if !ok {
			return nil, ErrInvalidFrame
		}
		return Bye{Reason: reason}, nil
	default:
		return nil, ErrInvalidFrame
	}
}

func EncodeCommand(command Command) ([]byte, error) {
	switch v := command.(type) {
	case Auth:
		if !utf8.ValidString(v.Code) {
			return nil, ErrInvalidFrame
		}
		return json.Marshal(struct {
			T    string `json:"t"`
			Code string `json:"code"`
		}{"auth", v.Code})
	case Move:
		if err := finite(v.Dx, v.Dy); err != nil {
			return nil, err
		}
		return json.Marshal(struct {
			T  string  `json:"t"`
			Dx float64 `json:"dx"`
			Dy float64 `json:"dy"`
		}{"move", v.Dx, v.Dy})
	case Scroll:
		if err := finite(v.Dx, v.Dy); err != nil {
			return nil, err
		}
		return json.Marshal(struct {
			T  string  `json:"t"`
			Dx float64 `json:"dx"`
			Dy float64 `json:"dy"`
		}{"scroll", v.Dx, v.Dy})
	case Click:
		if v.Button != ButtonLeft && v.Button != ButtonRight {
			return nil, ErrInvalidFrame
		}
		return json.Marshal(struct {
			T string `json:"t"`
			B string `json:"b"`
		}{"click", string(v.Button)})
	case KeyPress:
		if v.Key != KeyLeft && v.Key != KeyRight && v.Key != KeyEsc {
			return nil, ErrInvalidFrame
		}
		return json.Marshal(struct {
			T string `json:"t"`
			K string `json:"k"`
		}{"key", string(v.Key)})
	case ChordPress:
		if v.Chord != SpaceLeft && v.Chord != SpaceRight && v.Chord != MissionCtrl {
			return nil, ErrInvalidFrame
		}
		return json.Marshal(struct {
			T string `json:"t"`
			K string `json:"k"`
		}{"chord", string(v.Chord)})
	default:
		return nil, fmt.Errorf("unknown command")
	}
}

func EncodeReply(reply Reply) ([]byte, error) {
	switch v := reply.(type) {
	case Status:
		return json.Marshal(struct {
			T string `json:"t"`
		}{"status"})
	case Bye:
		if !utf8.ValidString(v.Reason) {
			return nil, ErrInvalidFrame
		}
		return json.Marshal(struct {
			T      string `json:"t"`
			Reason string `json:"reason"`
		}{"bye", v.Reason})
	default:
		return nil, fmt.Errorf("unknown reply")
	}
}

func MoveInRange(dx, dy float64) bool {
	return math.Abs(dx) <= MaxAbsMove && math.Abs(dy) <= MaxAbsMove
}

func number(value any) (float64, error) {
	n, ok := value.(json.Number)
	if !ok {
		return 0, ErrInvalidFrame
	}
	text := strings.TrimSpace(n.String())
	if text == "NaN" || text == "Infinity" || text == "-Infinity" {
		return 0, ErrNonfinite
	}
	f, err := n.Float64()
	if err != nil || math.IsInf(f, 0) || math.IsNaN(f) {
		return 0, ErrNonfinite
	}
	return f, nil
}

func finite(dx, dy float64) error {
	if math.IsInf(dx, 0) || math.IsNaN(dx) || math.IsInf(dy, 0) || math.IsNaN(dy) {
		return ErrNonfinite
	}
	return nil
}

func parseObject(data []byte) (map[string]any, error) {
	if !utf8.Valid(data) || !validUnicodeEscapes(data) {
		return nil, ErrInvalidFrame
	}
	dec := json.NewDecoder(bytes.NewReader(data))
	dec.UseNumber()
	opening, err := dec.Token()
	if err != nil || opening != json.Delim('{') {
		return nil, ErrInvalidFrame
	}
	raw := make(map[string]any)
	for dec.More() {
		token, err := dec.Token()
		if err != nil {
			return nil, ErrInvalidFrame
		}
		key, ok := token.(string)
		if _, duplicate := raw[key]; !ok || duplicate {
			return nil, ErrInvalidFrame
		}
		var value any
		if err := dec.Decode(&value); err != nil {
			return nil, ErrInvalidFrame
		}
		raw[key] = value
	}
	if closing, err := dec.Token(); err != nil || closing != json.Delim('}') {
		return nil, ErrInvalidFrame
	}
	var extra any
	if err := dec.Decode(&extra); err != io.EOF {
		return nil, ErrInvalidFrame
	}
	return raw, nil
}

func hasFields(raw map[string]any, fields []string) bool {
	if len(fields) == 0 || len(raw) != len(fields) {
		return false
	}
	for _, field := range fields {
		if _, ok := raw[field]; !ok {
			return false
		}
	}
	return true
}

// encoding/json replaces isolated UTF-16 surrogates; Swift rejects them.
func validUnicodeEscapes(data []byte) bool {
	quoted := false
	for i := 0; i < len(data); i++ {
		if data[i] == '"' {
			quoted = !quoted
			continue
		}
		if !quoted || data[i] != '\\' {
			continue
		}
		i++
		if i >= len(data) {
			return false
		}
		if data[i] != 'u' {
			continue
		}
		value, ok := unicodeEscape(data[i:])
		if !ok {
			return false
		}
		i += 4
		if value >= 0xdc00 && value <= 0xdfff {
			return false
		}
		if value >= 0xd800 && value <= 0xdbff {
			if i+2 >= len(data) || data[i+1] != '\\' {
				return false
			}
			low, ok := unicodeEscape(data[i+2:])
			if !ok || low < 0xdc00 || low > 0xdfff {
				return false
			}
			i += 6
		}
	}
	return true
}

func unicodeEscape(data []byte) (uint16, bool) {
	if len(data) < 5 || data[0] != 'u' {
		return 0, false
	}
	var value uint16
	for _, digit := range data[1:5] {
		value <<= 4
		switch {
		case digit >= '0' && digit <= '9':
			value |= uint16(digit - '0')
		case digit >= 'a' && digit <= 'f':
			value |= uint16(digit - 'a' + 10)
		case digit >= 'A' && digit <= 'F':
			value |= uint16(digit - 'A' + 10)
		default:
			return 0, false
		}
	}
	return value, true
}
